### PR TITLE
Wpilib24t4est

### PR DESCRIPTION
**Pull request by TurtleMinecraft**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -24,13 +24,18 @@ repositories {
...
+    implementation group: 'edu.wpi.first.wpilibNewCommands', name: 'wpilibNewCommands-java', version: '2024.1.1'
+    implementation group: 'edu.wpi.first.wpilibj', name: 'wpilibj-java', version: '2024.1.1'
+    implementation group: 'edu.wpi.first.cameraserver', name: 'cameraserver-java', version: '2024.1.1'
+    implementation group: 'edu.wpi.first.wpiutil', name: 'wpiutil-java', version: '2024.1.1'
+    implementation group: 'edu.wpi.first.wpimath', name: 'wpimath-java', version: '2024.1.1'
+    implementation group: 'edu.wpi.first.wpiunits', name: 'wpiunits-java', version: '2024.1.1'
+    implementation group: 'com.revrobotics.frc', name: 'REVLib-java', version: '2024.2.0'
+}
+
+dependencies {
```
Comment (Yoel-Perman-Brilliant) [inline build.gradle:38]: MERGE

File: src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java [hunk 2]
```diff
@@ -26,9 +26,9 @@ public class MotoredGenericSubsystem extends GenericSubsystem {
      * @param motorControllers the motor controllers in the subsystem
      */
     public MotoredGenericSubsystem(String namespaceName, Supplier<Double> minSpeed, Supplier<Double> maxSpeed,
-                                   MotorController... motorControllers) {
+                                   ArrayList<MotorController> motorControllers) {
         super(namespaceName, minSpeed, maxSpeed);
-        this.motorControllerGroup = new MotorControllerGroup(motorControllers);
+        this.motorControllerGroup = new ArrayList<>(motorControllers);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java:31]: List.of()

File: src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java:1]: potentially make documentations more coherent

File: src/main/java/com/spikes2212/command/genericsubsystem/MotorControllerGroup.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package com.spikes2212.command.genericsubsystem;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/genericsubsystem/MotorControllerGroup.java:1]: move to util

File: src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java [hunk 4]
```diff
@@ -14,7 +14,7 @@
  */
 public class MotoredGenericSubsystem extends GenericSubsystem {
 
-    protected MotorControllerGroup motorControllerGroup;
+    protected List<MotorController> motorControllers;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java:17]: if we add back motorControllerGroup, this seems like a pointless change

File: src/main/java/com/spikes2212/state/LayoutManager.java [hunk 5]
```diff
@@ -16,7 +16,7 @@ public void addLayout(T mode, ButtonLayout<K> layout) {
     }
 
     public Command getCommandFor(K button) {
-        return new SelectCommand(() -> allLayouts.get(currentLayout).getCommandForButton(button));
+        return new ProxyCommand(allLayouts.get(currentLayout).getCommandForButton(button));
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/state/LayoutManager.java:19]: lambda

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java [hunk 6]
```diff
@@ -14,7 +15,7 @@
  */
 public class MotoredGenericSubsystem extends GenericSubsystem {
 
-    protected MotorControllerGroup motorControllerGroup;
+    protected MotorControllerGroup motorControllers;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java:18]: change type to `MotorController`

File: src/main/java/com/spikes2212/util/MotorControllerGroup.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public void setVoltage(double outputVolts) {
+        for (MotorController motorController : motorControllers) {
+            motorController.setVoltage(inverted ? -outputVolts : outputVolts);
+        }
+    }
+
+    @Override
+    public double get() {
+        return 0.0;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/com/spikes2212/util/MotorControllerGroup.java:40]: ???

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java [hunk 8]
```diff
@@ -14,7 +15,7 @@
  */
 public class MotoredGenericSubsystem extends GenericSubsystem {
 
-    protected MotorControllerGroup motorControllerGroup;
+    protected MotorController motorControllers;
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/genericsubsystem/MotoredGenericSubsystem.java:18]: motorController

File: src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    public void configPIDF(PIDSettings pidSettings, FeedForwardSettings feedForwardSettings) {
+        TalonFXConfiguration config = new TalonFXConfiguration();
+        config.Slot0.kP = pidSettings.getkP();
+        config.Slot0.kI = pidSettings.getkI();
+        config.Slot0.kD = pidSettings.getkD();
+        config.Slot0.kS = feedForwardSettings.getkS();
+        config.Slot0.kV = feedForwardSettings.getkV();
+        config.Slot0.kA = feedForwardSettings.getkA();
+        config.Slot0.kG = feedForwardSettings.getkG();
+
```
Comment (Spikes-2212) [inline src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java:74]: remove

File: src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java [hunk 10]
```diff
@@ -23,46 +23,46 @@
 public class SparkMaxTankDrivetrain extends TankDrivetrain implements SmartMotorControllerTankDrivetrain {
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/drivetrains/smartmotorcontrollerdrivetrain/SparkMaxTankDrivetrain.java:23]: add ways to construct instances without providing a namespace name

File: src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+    /**
+     * The TalonFX which runs the loops.
+     */
+    protected final TalonFX master;
+
+    /**
+     * Additional TalonFXs that follow the master.
+     */
+    protected final List<? extends TalonFX> slaves;
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java:41]: why the wildcard type variable?

File: src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * Adds any data or commands to the {@link NetworkTable}s, which can be accessed using the {@link Shuffleboard}.
+     */
+    @Override
+    public void configureDashboard() {
+    }
+
+    @Override
+    public void configPIDF(PIDSettings pidSettings, FeedForwardSettings feedForwardSettings) {
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java:65]: this method does not apply the configuration it constructs

File: src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    @Override
+    public void configureTrapezoid(TrapezoidProfileSettings settings) {
+        MotionMagicConfigs config = new MotionMagicConfigs();
+        config.MotionMagicAcceleration = settings.getAccelerationRate();
+        config.MotionMagicCruiseVelocity = settings.getMaxVelocity();
+        master.getConfigurator().apply(config);
+    }
+
+    @Override
+    public void configureLoop(PIDSettings pidSettings, FeedForwardSettings feedForwardSettings,
```
Comment (TurtleMinecraft) [inline src/main/java/com/spikes2212/command/genericsubsystem/smartmotorcontrollersubsystem/TalonFXGenericSubsystem.java:85]: neither does this one

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

**Review by Yoel-Perman-Brilliant — APPROVED** [review]

**Review by Spikes-2212 — APPROVED** [review]

**Pull request merged**